### DISCUSSION TITLE
Download CPM via CMake

### DISCUSSION THREAD

**Discussion by Leon0402** [cpm-cmake/CPM.cmake#320] [Q&A]
According to: https://github.com/cpm-cmake/CPM.cmake/wiki/Downloading-CPM.cmake-in-CMake

> Note that downloading CPM.cmake using the script above will not allow you to create new builds of your project when offline, even if all dependencies are available in the CPM_SOURCE_CACHE. Therefore, especially for libraries, it is recommended to add the CPM.cmake script to the source repository.

Is the reason for this just because the file(DOWNLOAD) is always done? Or needs to be done at least one time for each new build directory? 

I currently use CPM via FetchContent, which works fine. For each new build directory internet is needed as well, but this seems okay to me. 
The problem is only that I get a warning that I'm not using a release version even though I am. It seems the tagged release version doesn't contain the right version info, but only the released version in the assets. 

I think a tagged release version such as https://github.com/cpm-cmake/CPM.cmake/tree/v0.34.3 should also contain the right version. What do you think?

I like to use FetchContent as it allows me to specify any Commit hash, no matter if it's a release or just the current master commit. This makes it very easy to switch between different versions. It's also more consistent with the CPM code compared to the file(DOWNLOAD) approach.

**Comment by TheLartians** [comment]
> Is the reason for this just because the file(DOWNLOAD) is always done? Or needs to be done at least one time for each new build directory?

Actually I think that paragraph is outdated, as the snippet checks the cache directory as well. I've removed the note from the wiki. Would the script be a better option for you in this case?

> For each new build directory internet is needed as well, but this seems okay to me.

There are still quite a bit of use-cases that require offline compatibility, e.g. for security reasons. Note that it would also forbid any users of your project to build offline as well.

> I think a tagged release version such as https://github.com/cpm-cmake/CPM.cmake/tree/v0.34.3 should also contain the right version. What do you think?

The release assets are currently [generated automatically](https://github.com/cpm-cmake/CPM.cmake/blob/master/.github/workflows/publish.yaml) whenever there's a new tag. I see how adding the version to master would be nice, however adding the according release version to the main branch would come with a maintenance overhead for each release that I would like to avoid.

**Reply by Leon0402** [reply]
I do think that using the script is an option, although I still prefer FetchContent as it is simpler and allows to change the GIT_TAG to any git commit very easy. 

> > 
> 
> There are still quite a bit of use-cases that require offline compatibility, e.g. for security reasons. Note that it would also forbid any users of your project to build offline as well.

Which would be possible to avoid with FetchContent as well. You can provide a path to a local directory. So for such special circumstances, there is a solution. 



> The release assets are currently [generated automatically](https://github.com/cpm-cmake/CPM.cmake/blob/master/.github/workflows/publish.yaml) whenever there's a new tag. I see how adding the version to master would be nice, however adding the according release version to the main branch would come with a maintenance overhead for each release that I would like to avoid.

I see. Maybe it would be possible to trigger a commit from the pipeline?

**Reply by TheLartians** [reply]
> I see. Maybe it would be possible to trigger a commit from the pipeline?

Perhaps, though I'm not sure how to handle it nicely atm, as we would need to set and then reset the git tag. Alternatively, we could also add a zipped release artefact, that you could easily download using `FetchContent` as well. Would that work for you as well?

E.g. something like this could be possible:

```cmake
FetchContent_Declare(
  CPM.cmake
  URL https://github.com/cpm-cmake/CPM.cmake/releases/download/v0.34.3/CPM.cmake.zip
)
```

**Comment by Leon0402** [comment]
I saw in another discussion that the release version also removes some bloat? Is this really true? I cannot see anything in the pipeline except the version changing?

**Reply by TheLartians** [reply]
I think what was referred to was the [`get_cpm.cmake`](https://github.com/cpm-cmake/CPM.cmake/releases/download/v0.34.3/get_cpm.cmake) script, that is published alongside the full script. This will fetch CPM.cmake on-demand (and also work offline later), avoiding the having to add the full script to your project repos. Of course if you already use `FetchContent` or `file(DOWNLOAD)` it won't make a big difference in that regard.
